### ISSUE TITLE
MPS] torch.nanmedian SIGABRT on ndim > 4 with dim=0 (MPSNDArraySort axis limit)

### ISSUE THREAD

**Issue by christianahmed905** [pytorch/pytorch#187016]
### 🐛 Describe the bug

"""
torch.nanmedian — SIGABRT (process crash) on MPS for high-rank inputs

torch.nanmedian crashes the Python process with SIGABRT (exit code -6) on MPS
when the input tensor has ndim > 4 and the reduction axis maps to position
>= 4 from the right (e.g. dim=0 on a 5D tensor).  No Python exception is
raised; the process is killed and cannot be recovered.

Crash origin:
  MPSNDArraySort.mm:252: failed assertion `(null)' Axis = 4.
    This class only supports axis = 0, 1, 2, 3

Root cause: torch.nanmedian on MPS uses the same MPSNDArraySort kernel as
torch.sort.  That kernel right-aligns dimensions and only handles the
rightmost 4 axes.  PyTorch does not validate this constraint before
dispatching to MPS, so high-rank tensors with dim in the "left" portion
trigger a fatal assertion inside the Metal runtime.

General crash rule:  crash when  dim < ndim - 4  (i.e. axis from right >= 4)

Failing example (fuzz crash):
  shape=(2, 1, 3, 1, 2)  rank=5  dim=0  keepdim=True  dtype=int64

Crash statistics (Specialized generator, mps):
  torch.nanmedian — 1 crash (hash [PROC], signal=6)
                    example: shape=(2, 1, 3, 1, 2) dim=0 keepdim=True
                             dtype=int64 has_nan=False

Expected: (values, indices) tensors, OR a catchable RuntimeError (e.g.
          "MPS supports at most 4 dimensions for this operation"), or CPU
          fallback.
Actual:   SIGABRT kills the Python process — not catchable with try/except.

Related:  same MPSNDArraySort limit affects torch.sort / torch.argsort
          (see mps_high_rank_sigabrt_20260515.py).  torch.median may also
          be affected.
Affected op:     torch.nanmedian (dim-form reduction)
Affected device: mps (Apple Silicon)
Not affected:    ndim <= 4 on MPS; same inputs on CPU
Versions:        PyTorch 2.12.0, macOS 26.4.1 arm64 (Apple Silicon MPS)
"""

import torch

print(f"torch version: {torch.__version__}")
assert torch.backends.mps.is_available(), "MPS not available — run on Apple Silicon"

device = torch.device("mps")

#  -- nanmedian ndim=5 dim=0 int64 (fuzz example) --
x5 = torch.randint(0, 10, (2, 1, 3, 1, 2), dtype=torch.int64, device=device)
print(f"input shape: {tuple(x5.shape)}  dim=0  keepdim=True  dtype={x5.dtype}")
print("torch.nanmedian(x5, dim=0, keepdim=True) → SIGABRT")
torch.nanmedian(x5, dim=0, keepdim=True)

print("(no crash triggered — uncomment a block above to reproduce)")

print("\n[ref] CPU (same inputs, should succeed):")
x5_cpu = torch.randint(0, 10, (2, 1, 3, 1, 2), dtype=torch.int64)
vals, idx = torch.nanmedian(x5_cpu, dim=0, keepdim=True)
print(f"  values shape: {tuple(vals.shape)}  indices shape: {tuple(idx.shape)}")

### Versions

PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.5 (main, May 10 2026, 19:20:57) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-26.5-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[conda] Could not collect

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20
